### ISSUE TITLE
Connection slows over time

### ISSUE THREAD

**Issue by fonty422** [tiangolo/fastapi#2760]
### Example

This is cut back - I'm not including various api calls to endpoints.

```Python
# server.py
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse, HTMLResponse, FileResponse
from fastapi.middleware.cors import CORSMiddleware
from starlette.websockets import WebSocket, WebSocketDisconnect
from asyncio import sleep, ensure_future

app = FastAPI(title="App")

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

template="../../build/index.html"

@app.get("/")
def default_html():
  with open(template, 'r') as f:
    response=f.read()
    
  headers = {"Cache-Control": "max-age=31536000"}
  return HTMLResponse(content=response, headers=headers)
```
Then run with:

```
uvicorn server:app --host = 0.0.0.0 --port = 80
```

### Description

* Open the browser and call the endpoint `/`.
* Should return the html and it does.
* Over time it appears to take longer and longer for the page to display on the browser.
* trying to download the file directly (the apparent slow js file) from /static/js/<filename>.chunk.js displays text that appears to download in pieces.

### Environment

* OS: Centos 8
* FastAPI Version: 0.54.1
* Python version: 3.8.2

### Additional context
I've built a ReactJS front end with FastAPI as the back end.
The back end performs 3 tasks:
1. returns html to display
2. handles http requests from clients
3. sends server-initiated messages to clients

When I first run the line to start the `uvicorn` service, the response times are nice and quick - less than 2 seconds the page is fully rendered on the browser.
Over time, it takes longer and longer to render the page. A few days later it can take up to 15 seconds. We did see an 80 second load time once before we restarted hte unvicorn service. What we notice when we try to inspect what is taking so long, it's the downloading of a 2MB javascript file, which is referenced in the html file as a `<script />`. If we head to the IP of that file directly it it appears to strongly mimic the time to display the page, and we literally watch lines of code download one at a time.
Perhaps fair if we're grabbing something large, but it's quite small and also hosted on the same network we're experiencing the slow speed.

**Comment by falkben** [comment]
Please use the issue template and include example code if you can.

**Comment by fonty422** [comment]
Should I be using templating:

**Comment by ycd** [comment]
@fonty422 you are using a constant HTML file, why you are not caching it. 

Load the HTML in the memory for once, then use it.

**Comment by fonty422** [comment]
Ok, so at the outset when initialting the app, first load the html content in to a variable object then ensure I send that object back as the response?
I'm still not sure, though, why it takes progressively longer for the full response to be received.

**Comment by aalifadv** [comment]
I think what @ycd was referring to is that you seem to read the template file each time a GET request comes in to the "/" endpoint. You should not, assuming the file is static and doesnt change. 

You can do something like this:

```python
import functools

@functools.lru_cache
def get_and_cache_file_content(file_path):
    with open(file_path) as file_obj:
        return file_obj.read()
```
Then change your app.get to be something like: 

```python
@app.get("/")
def default_html():
  response=get_and_cache_file_content(template)
  headers = {"Cache-Control": "max-age=31536000"}
  return HTMLResponse(content=response, headers=headers) 
```

The first time a request comes in, it will read template from the desk, next time it will not.

finally, it is hard to debug a performance issue with the limited information provided. 

> it's the downloading of a 2MB javascript file, which is referenced in the html file as a <script />.

Who is serving  that file? is it external or it is fastAPI? 
What is the size of the file that is referenced by template variable?

**Comment by fonty422** [comment]
Thanks @aalifadv. I tested out your method and it now takes ~30 seconds to load the page every time.
I've attached an image showing the load stats. Nothing there is by any means large, but some things seem to take an age to download.
![download_time](https://user-images.githubusercontent.com/18232646/108422956-53fc5b00-728b-11eb-8c35-064bc644ac6c.jpg)
I did also move the read file to outside all functions  and save the html as a string, and the results were as previously described. It loads the page quickly straight after the server is started, then it progressively becomes slower the the longer the server uptime.
The template file is only 2kb.

**Comment by fonty422** [comment]
So there's still no solution here. Note that the chunk.js file is in the header of the html file that is sent back. So my, possibly naïve, understanding is that the html file is returned which contains no information on what to display, the header contains this js file, which _are_ the instructions on what to render to the page. THIS is what is slow to download. And it makes no sense why. Again if i restart the fastapi server, it's quick, but as the days go on it becomes slower and slower until it's so painfully slow that it just has to be restarted and it's really annoying the users.

Any thoughts or assistance would be greatly appreciated!

**Comment by fonty422** [comment]
I've just implemented code splitting with lazy loading on the React side which results in there being 28 chunk files rather than 2, with the smallest being 1.1kB and the largest being 282kB, while the average is only about 40kB. Hopefully this might make things better, but if it does I have no idea why it works. unless there's a file size setting where below this is serves directly and above this it streams it in pieces of some default size.

**Comment by ycd** [comment]
Have you done any profiling to find where bottleneck comes from?

**Comment by fonty422** [comment]
No. I just noticed that i'm running an older version of React too, so the standard react profiling tool doesn't appear to work.
But just running the performance test in chrome yields the following.
![image](https://user-images.githubusercontent.com/18232646/112083511-48c68300-8bdb-11eb-8fee-fd7c91c49316.png)

Again, the hang-up appears to be the time to download. I can watch these files (which are now small with the largest being in the order of 150kB) slowly download in real time. This one, which is 162kB took 8.81s to download a week after I restarted the server. 
![image](https://user-images.githubusercontent.com/18232646/112084008-2ed97000-8bdc-11eb-93cd-4ffa8afa2be3.png)

And again, the strange behaviour is that if i restart the server, then the total load time is 2-3 seconds, a week later it's 25s, then a week later something like a minute.

To be honest, i'm tempted to try flask to serve the page and fastapi for all other api and socket requests.

**Comment by falkben** [comment]
You should try py-spy. You can attach to a running process.

**Comment by fonty422** [comment]
Thanks for the advice @falkben. I ran it with the top command and noted the following.

![Screenshot from 2021-03-23 17-03-48](https://user-images.githubusercontent.com/18232646/112101932-981cab80-8bfb-11eb-8e8c-33fc017831f4.png)

* The GIL is almost at 100% and often hits 100%.
* The Active is always significantly over 100% and often reaches ~300%.
 
Are these things an issue?

**Comment by fonty422** [comment]
So i decided to just try run it with Flask to serve the HTML and FastAPI to handle all web socket and API calls, and it's been running strong, fast, and with no reduction in performance for over 24 hours now. In the past, even at 1 day, we'd see performance drops.
I'm happy to leave it at this, but if anyone could shed light on why it might be happening in the first instance, that would be useful.

**Comment by hdsouza06** [comment]
I am facing similar issue , the Fast API url running through postman is very fast , however when we integrate the API with front end application written in next JS it is taking lot of time .
For an average , running it directly through postman takes me about 500 ms as compared to running it from the front end application where it is taking almost 5 to 9 seconds .

what might be the possible reason for this .

your comments or feedbacks are highly appreciated
